src/algorithms/l0_smooth.jl

```diff
@@ -0,0 +1,51 @@
+struct L0Smooth <: AbstractImageSmoothAlgorithm
+    """smoothing weight"""
+    λ::Float64
+    """iteration rate"""
+    κ::Float64
+
+    function L0Smooth(λ, κ)
+        λ > zero(λ) || throw(ArgumentError("smoothing weight λ should be positive"))
+        κ > one(κ) || throw(ArgumentError("iteration rate κ should be > 1.0"))
+        new(λ, κ)
+    end
+end
+
+L0Smooth(; λ::Float64=2e-2, κ::Float64=2.0) = L0Smooth(λ, κ)
+
+function (f::L0Smooth)(out::GenericGrayImage,
+                       img::GenericGrayImage)
+    S = float.(channelview(img))
+    λ = f.λ
+    κ = f.κ
+    βmax = 1e5
+    fx = [1 -1]
+    fy = [1, -1]
+    N, M = size(S)
+    sizeI2D = (N, M)
+    sizeI2D_t = (M, N)
+    otfFx = freqkernel(centered(fx), sizeI2D)
+    otfFy = transpose(freqkernel(centered(transpose(fy)), sizeI2D_t))
+    Normin1 = fft(S, (1, 2))
+    Denormin2 = abs.(otfFx).^2 + abs.(otfFy).^2

```

Comment:
`abs.(otfFx).^2` and `abs.(otfFy).^2` are two unnecessary intermediate memory allocations. This can be eliminated by broadcasting to all operations:

```suggestion
    Denormin2 = @. abs(otfFx)^2 + abs(otfFy)^2
```

```julia
f(x, y) = abs.(x).^2 + abs.(y).^2
g(x, y) = abs.(x).^2 .+ abs.(y).^2
h(x, y) = @. abs(x)^2 + abs(y)^2

x = rand(64, 64);
y = rand(64, 64);

@btime f($x, $y); # 4.197 μs (6 allocations: 96.23 KiB)
@btime g($x, $y); # 1.334 μs (2 allocations: 32.08 KiB)
@btime h($x, $y); # 1.379 μs (2 allocations: 32.08 KiB)
```

See also the discussion in Chinese discourse https://discourse.juliacn.com/t/topic/3934